File: fortran/perf/perf_pathological.f90
! perf_pathological.f90 — demonstrates the two scaling bottlenecks
!
! Construction: add_command writes into a capacity-doubling over-allocated
! commands(:) array (initial capacity 8, doubling via move_alloc when full).
! finalize trims each state to the exact count in one reallocation (O(N) total).
!
! Dispatch: find_matches must scan the entire command list of the current state
! to detect ambiguity, performing three allocatable character assignments per
! entry (req, opt, full).  Every dispatch is therefore O(N) heap allocations
! regardless of which command was typed.
!
! For interactive use these costs are immaterial -- graphs have tens of commands
! per state and dispatch happens at human speed.  This program makes them
! visible by scaling N to the point where the time becomes measurable, then
! shows the same workload on a realistic-sized graph.
!
! Build and run:  make -C .. F=ifx perf
!            or:  make -C .. F=gfortran perf

module perf_procs
    use cmdgraph
    use dlist
    implicit none
contains
    function act_noop(args, ctx) result(rv)
        type(dlist_t), intent(in)    :: args
        character(len=*), intent(in) :: ctx
        type(action_result_t)        :: rv
    end function act_noop

    subroutine build_engine(eng, n)
        type(engine_t), intent(out) :: eng
        integer,        intent(in)  :: n
        integer          :: i
        character(len=7) :: spec     ! "c" + 5-digit zero-padded number = 6 chars
        call eng%add_state("root", prompt="> ")
        do i = 1, n
            write(spec, '("c",i5.5)') i    ! c00001, c00002, ..., cNNNNN
            call eng%add_command("root", trim(spec), EDGE_ACTION, proc=act_noop)
        end do
        call eng%finalize("root")
        call eng%set_io_units(output_unit=QUIET_UNIT, error_unit=QUIET_UNIT)
    end subroutine build_engine
end module perf_procs

program perf_pathological
    use cmdgraph
    use perf_procs
    implicit none

    ! --- Tune to taste ---
    ! At N=5000, D=100000 on a modern desktop (ifx -O3) the pathological
    ! dispatch phase takes ~13 s; the realistic phase takes < 0.1 s.
    integer, parameter :: N_BIG      = 5000    ! pathological: commands per state
    integer, parameter :: N_SMALL    = 20      ! realistic:    commands per state
    integer, parameter :: D          = 100000  ! dispatches per timing phase

    type(engine_t) :: eng
    integer        :: i, rc
    real           :: t0, t1

    write(*, '(/,a,/)') "=== pathological case (N=" // itoa(N_BIG) // ") ==="

    ! -- Construction O(N) --
    ! add_command writes into a capacity-doubling commands(:) array.
    ! finalize trims each state to the exact count: O(N) total.
    call cpu_time(t0)
    call build_engine(eng, N_BIG)
    call cpu_time(t1)
    write(*, '(a,f7.3,a)') "  construction (add_command x " // itoa(N_BIG) // "):  ", &
        t1 - t0, " s   [O(N)]"

    ! -- Dispatch O(N) per call --
    ! find_matches scans all N entries every call; note that "best" and "worst"
    ! command positions make no difference -- there is no early exit because the
    ! full scan is needed for ambiguity detection.
    call cpu_time(t0)
    do i = 1, D
        rc = eng%dispatch("c00001")   ! first command in the list
    end do
    call cpu_time(t1)
    write(*, '(a,f7.3,a)') "  dispatch x " // itoa(D) // ":                        ", &
        t1 - t0, " s   [O(N x D)]"

    write(*, '(/,a,/)') "=== realistic case (N=" // itoa(N_SMALL) // ") ==="

    call cpu_time(t0)
    call build_engine(eng, N_SMALL)
    call cpu_time(t1)
    write(*, '(a,f7.3,a)') "  construction (add_command x " // itoa(N_SMALL) // "):   ", &
        t1 - t0, " s"

    call cpu_time(t0)
    do i = 1, D
        rc = eng%dispatch("c00001")
    end do
    call cpu_time(t1)
    write(*, '(a,f7.3,a)') "  dispatch x " // itoa(D) // ":                        ", &
        t1 - t0, " s"

contains
    pure function itoa(n) result(s)
        integer, intent(in)           :: n
        character(len=:), allocatable :: s
        character(len=20)             :: buf
        write(buf, '(i0)') n
        s = trim(buf)
    end function itoa

end program perf_pathological

File: cpp/perf/perf_pathological.cxx
// perf_pathological.cxx — demonstrates the two scaling bottlenecks
//
// Construction: add_command appends to an internal std::vector<Command>
// which grows via the standard capacity-doubling strategy; finalize copies
// to the resolved commands vector in O(N) total.
//
// Dispatch: find_matches must scan the entire command list of the current state
// to detect ambiguity.  Every dispatch is therefore O(N) regardless of which
// command was typed.
//
// For interactive use these costs are immaterial — graphs have tens of commands
// per state and dispatch happens at human speed.  This program makes them
// visible by scaling N to the point where the time becomes measurable, then
// shows the same workload on a realistic-sized graph.
//
// Build and run:  make -C .. CXX=icpx perf
//            or:  make -C .. CXX=g++  perf

#include "cmdgraph.hxx"
#include <cstdio>
#include <cstring>
#include <time.h>

static cmdgraph::ActionResult act_noop(const cmdgraph::ArgList&, const std::string&)
{
    return cmdgraph::action_ok();
}

static void build_engine(cmdgraph::Engine& eng, int n)
{
    eng = cmdgraph::Engine{};
    eng.add_state("root", "> ");
    char spec[8];
    for (int i = 1; i <= n; ++i) {
        std::snprintf(spec, sizeof(spec), "c%05d", i);
        eng.add_command("root", spec, cmdgraph::EdgeKind::Action,
                        {.proc = act_noop});
    }
    eng.finalize("root");
    eng.set_io(nullptr, nullptr, nullptr);
}

static double elapsed_s(const struct timespec& a, const struct timespec& b)
{
    return (b.tv_sec - a.tv_sec) + (b.tv_nsec - a.tv_nsec) * 1e-9;
}

static std::string itoa(int n)
{
    char buf[24];
    std::snprintf(buf, sizeof(buf), "%d", n);
    return buf;
}

int main()
{
    constexpr int N_BIG   = 5000;
    constexpr int N_SMALL = 20;
    constexpr int D       = 100000;

    cmdgraph::Engine eng;
    struct timespec t0, t1;

    std::printf("\n=== pathological case (N=%s) ===\n\n", itoa(N_BIG).c_str());

    // -- Construction O(N) amortised --
    clock_gettime(CLOCK_MONOTONIC, &t0);
    build_engine(eng, N_BIG);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    std::printf("  construction (add_command x %s):  %7.3f s   [O(N) amortised]\n",
                itoa(N_BIG).c_str(), elapsed_s(t0, t1));

    // -- Dispatch O(N) per call --
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < D; ++i)
        (void)eng.dispatch("c00001");
    clock_gettime(CLOCK_MONOTONIC, &t1);
    std::printf("  dispatch x %s:                        %7.3f s   [O(N x D)]\n",
                itoa(D).c_str(), elapsed_s(t0, t1));

    std::printf("\n=== realistic case (N=%s) ===\n\n", itoa(N_SMALL).c_str());

    clock_gettime(CLOCK_MONOTONIC, &t0);
    build_engine(eng, N_SMALL);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    std::printf("  construction (add_command x %s):   %7.3f s\n",
                itoa(N_SMALL).c_str(), elapsed_s(t0, t1));

    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < D; ++i)
        (void)eng.dispatch("c00001");
    clock_gettime(CLOCK_MONOTONIC, &t1);
    std::printf("  dispatch x %s:                        %7.3f s\n",
                itoa(D).c_str(), elapsed_s(t0, t1));

    return 0;
}
